### PR TITLE
Remove tvars introduced while testing normalizedCompatible

### PR DESCRIPTION
**Pull request by dwijnand**
Having looked at the type var constraints for so long, I wanted to finally get rid of these temporary typevars (and a type lambda) added for the purpose of typing a qualifier against a selection prototype, and when the selection is actually typed, brand new typevars (and a type lambda) are added.  Given it's from the same type lambda, they're necessarily all called the same, so they're indistinguishable in the constraint debugging.

### CODE REVIEW

**Comment by dwijnand** [comment]
@odersky I assume you've seen this as well.  I'm not sure of the best way to fix this, but this way works and seems clean enough.  But I'm happy to experiment with other ways, if you have ideas.

**Comment by odersky** [comment]
I am out of bandwidth to page in the context for this one. Maybe @smarter can take a look?

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 1]
```diff
@@ -71,6 +71,10 @@ object ProtoTypes {
                    |constraint was: ${ctx.typerState.constraint}
                    |constraint now: ${newctx.typerState.constraint}""")
             if result && (ctx.typerState.constraint ne newctx.typerState.constraint) then
+              val tvars = (newctx.typerState.ownedVars -- ctx.typerState.ownedVars).toList
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:74]: Isn't this equal to `newctx.typerState.ownedVars` ? Otherwise the assert in https://github.com/scala/scala3/blob/eb5c3e81304eee5ce75c2796bd4b7b16d2d1b19b/compiler/src/dotty/tools/dotc/core/TyperState.scala#L288-L289 should trigger

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 2]
```diff
@@ -71,6 +71,10 @@ object ProtoTypes {
...
+              inContext(newctx):
+                Inferencing.instantiateSelected(tp, tvars)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:76]: If the tvars are created from the TypeLambda underlying `tp`, then I expect that they wouldn't be present at all in `tp` so this line would be a no-op, is that not the case?

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 3]
```diff
@@ -71,6 +71,10 @@ object ProtoTypes {
...
+                for tvar <- tvars do if !tvar.isInstantiated then tvar.instantiate(fromBelow = false)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:77]: Does `fromBelow = false` rather than `true` make a difference here?

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala (hunk 1 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:74]: This was meant to be the shorthand for `newctx.typerState.ownedVars -- locked`, with a `val locked = newctx.typerState.ownedVars` before testCompat.  But I'll doublecheck what you're pointing to.

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala (hunk 2 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:76]: I'll check

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala (hunk 3 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:77]: No, it's just following the norm of maximising unconstrained invariant lambda parameters to Any rather than minimising to Nothing.  The goal is to instantiate them with anything, so the lambda can be `.remove`'d from the constraint.

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:77]: Used `Inferencing.instantiateSelected` + this, as `instantiateSelected` is the only current API where you can specify the tvars (and I didn't want to add yet more general purpose ways to instantiate things..)

Comment (smarter) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:77]: > No, it's just following the norm of maximising unconstrained invariant lambda parameters

Are we certain that they're unconstrained and not constrained by something in `pt`?

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala (hunk 1 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:74]: Yes, you're right, will simplify.

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala (hunk 2 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala]: You're right.  Wasn't looking at the right type.

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala (hunk 3 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala]: Changed the approach.

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 4]
```diff
@@ -71,7 +71,8 @@ object ProtoTypes {
                    |constraint was: ${ctx.typerState.constraint}
                    |constraint now: ${newctx.typerState.constraint}""")
             if result && (ctx.typerState.constraint ne newctx.typerState.constraint) then
-              newctx.typerState.commit()
+              newctx.typerState.gc()     // Remove any type lambdas and tvars added via testCompat
+              newctx.typerState.commit() // Commit the rest of the typer state information
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:75]: It's not clear to me that this makes a difference: `commit()` will call `gc()` on the target state after moving type lambdas to that state, so I'd expect anything that `newctx.typerState.gc()` does to be subsumed by that?

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala]: You were right again 😄.
So I slowed down and tried again.

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 5]
```diff
@@ -71,6 +71,14 @@ object ProtoTypes {
                    |constraint was: ${ctx.typerState.constraint}
                    |constraint now: ${newctx.typerState.constraint}""")
             if result && (ctx.typerState.constraint ne newctx.typerState.constraint) then
+              // Remove all type lambdas and tvars introduced by testCompat
+              for tvar <- newctx.typerState.ownedVars do
+                val tl = tvar.origin.binder
+                newctx.typerState.ownedVars -= tvar
+                if newctx.typerState.constraint.contains(tl) then
+                  newctx.typerState.constraint = newctx.typerState.constraint.remove(tl)(using newctx)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:79]: I think this is correct but a bit low-level and instead we could write:
```suggestion
              // Remove all tvars introduced by testCompat
              inContext(newctx):
                for tvar <- newctx.typerState.ownedVars do
                  if !tvar.isInstantiated then
                    tvar.instantiate(fromBelow = true)
```
and let `commit()` remove the type lambda.

File: compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala [hunk 6]
```diff
@@ -360,20 +360,21 @@ trait TypeAssigner {
...
+            wrongNumberOfTypeArgs(fn.tpe, pt.typeParams, args, tree.srcPos)
           else {
             // Make sure arguments don't contain the type `pt` itself.
-            // make a copy of the argument if that's the case.
+            // Make a copy of `pt` if that's the case.
             // This is done to compensate for the fact that normally every
             // reference to a polytype would have to be a fresh copy of that type,
             // but we want to avoid that because it would increase compilation cost.
             // See pos/i6682a.scala for a test case where the defensive copying matters.
-            val ensureFresh = new TypeMap with CaptureSet.IdempotentCaptRefMap:
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala:370]: How are the changes in this file and Typer related to the change in ProtoTypes?

**Review by smarter — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala (hunk 6 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala:370]: So the ones in Typer, i.e. when adapting by applying fresh type vars, is necessary for the recursive programs like i6682a (and in the comment).  I then randomly grepped the reference i6682a and found the code in TypeAssigner, so I started expanding my Typer comment.  But I noticed that the way its done in TypeAssigner is in cloning the lambda in the args.  So I thought I could hit two birds with one stone by (1) do it consistently in both (making it easier to document) and (2) do it more efficiently in TypeAssigner by doing it once rather than for each argument.

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala (hunk 5 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:79]: Sure

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 7]
```diff
@@ -500,29 +500,6 @@ object ProtoTypes {
...
           // TyperState and if that created additional constraints.
           if (passedTyperState ne protoTyperState) && (oldConstraint ne newConstraint) then
-            // To respect the pre-condition of `mergeConstraintWith` and keep
-            // `protoTyperState` committable we must ensure that it does not
-            // contain any type variable which don't already exist in the passed
-            // TyperState. This is achieved by instantiating any such type
-            // variable. NOTE: this does not suffice to discard type variables
-            // in ancestors of `protoTyperState`, if this situation ever
-            // comes up, an assertion in TyperState will trigger and this code
-            // will need to be generalized.
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:503]: I believe this code is flawed since the situation in the NOTE can happen which I think is what causes https://github.com/scala/scala3/issues/21555 and similar issues, but it's not clear to me that just removing this is fine, and it seems unrelated to the rest of this PR?

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala]: I went to reverify things, and I can't find a test case that fails without your code here (and without any thing in this PR).  So I can't say that my code replaces it, because it doesn't seem to do anything for any of our tests.  So I put it back (by dropping my commit).

**Review by smarter — APPROVED** [review]
